### DISCUSSION TITLE
Create multiple LXM Routers with different identities

### DISCUSSION THREAD

**Discussion by r8io** [markqvist/Reticulum#547] [Help & Questions] [answered]
### Question
How do I create multiple LXM Routers with different identities in a single application. My goal is to provide multiple LXMF destinations, created from different identities, for interaction with an application.

### My naive approach
My naive approach looks like this (simplified):
```python
 message_router1 = LXMF.LXMRouter(identity1, storagepath1)
 message_router2 = LXMF.LXMRouter(identity2, storagepath2)
```  

Which produces the following error:
```shell
Attempt to register an already registered destination.
```

`Versions used:  rns  0.7.7, lxmf  0.5.0`

### Disclaimer
I do not know anything about LXMF besides the examples Mark provided. My naive "solution" at the end is most likely not a good solution since I clearly wander in forbidden land here. I do not even know what a `lxmf.query` destination is used for :-)

There probably is a much more adequate way to go about this problem without the "need" to change RNS source code.

### Problem
It appears that every LXM Router creates three destinations:
- `lxmf.query`
- `lxmf.propagation`
- `lxmf.delivery`

The `lxmf.query` destination is created WITHOUT an identity, which results in the creation of identical `lxmf.query` destinations even if LXM Routers are created with different identities. This leads to the `Attempt to register an already registered destination` error mentioned above.

LXMF.LXMRouter.py (line: 120):
```python
self.lxmf_query_destination  = RNS.Destination(None, RNS.Destination.IN, RNS.Destination.PLAIN, APP_NAME, "query")
```        

### My naive "Solution"
The original code prevents the registration of existing destinations by raising an error. This prevents the creation of multiple LXM Routers, at least in the way I go about it as explained above.

RNS.Transport.py (line 1948):
```python
def register_destination(destination):
    destination.MTU = RNS.Reticulum.MTU
    if destination.direction == RNS.Destination.IN:
        for registered_destination in Transport.destinations:
            if destination.hash == registered_destination.hash:
                raise KeyError("Attempt to register an already registered destination.")

        Transport.destinations.append(destination)

        if Transport.owner.is_connected_to_shared_instance:
            if destination.type == RNS.Destination.SINGLE:
                destination.announce(path_response=True)
``` 

My naive "solution" prevents the registration of existing destinations by skipping. By doing so I can create multiple LXM Routers, but only ONE `lxmf.query` destination is created. On a superficial level that seems to work, but considering my disclaimer above the validity of this approach should be questioned ;-)
```python
def register_destination(destination):
        destination.MTU = RNS.Reticulum.MTU
        if destination.direction == RNS.Destination.IN:
            destination_exists = False
            for registered_destination in Transport.destinations:
                if destination.hash == registered_destination.hash:
                    # raise KeyError(f"RNS.Transport() Attempt to register an already registered destination. {destination}")  # r8io
                    destination_exists = True
                    RNS.log(f"RNS.Transport() Attempt to register an already registered destination. {destination}", level=2)
                    continue

            if not destination_exists:
                Transport.destinations.append(destination)
                RNS.log(f"RNS.Transport() Registered destination. {destination}")

                if Transport.owner.is_connected_to_shared_instance:
                    if destination.type == RNS.Destination.SINGLE:
                        destination.announce(path_response=True)
```

**Comment by markqvist** [comment, marked as answer]
Thanks for your detailed post about this! Fortunately, the real fix is very easy. The `query_destination` is a very old artifact on a very old version of LXMF, and it has not been used for anything for about three years. The fix is to remove it, as I have just done in commit [this commit](https://github.com/markqvist/LXMF/commit/453772859d4e806577839ac82a985e54ad15857d).

Thanks for spotting and investigating it!

You should be able to pull the latest source from the repo and do as you want now :)

**Comment by r8io** [comment]
Thank you for for your quick reply and the easy fix! I feel like the blind hen that found a kernel of grain ... relay happy :-) It works nicely now.
